File: lib/domain/weather_data/gateways/internal_weather_fetch_start_gateway.rb
# frozen_string_literal: true

module Domain
  module WeatherData
    module Gateways
      # 内部 API 用: 農場の天気データ取得ジョブ投入（永続化境界は Adapter）。
      module InternalWeatherFetchStartGateway
        WeatherFetchFarmSnapshot = Struct.new(:farm_id, :weather_data_status, :weather_data_count, :total_blocks, keyword_init: true)

        StartInternalWeatherFetchResult = Struct.new(:kind, :snapshot, :error_message, keyword_init: true) do
          def self.farm_not_found
            new(kind: :farm_not_found)
          end

          def self.completed(snapshot)
            new(kind: :completed, snapshot: snapshot)
          end

          def self.started(snapshot)
            new(kind: :started, snapshot: snapshot)
          end

          def self.failed(message)
            new(kind: :failed, error_message: message)
          end
        end

        def start_internal_weather_data_fetch(farm_id:)
          raise NotImplementedError, "#{self.class} must implement start_internal_weather_data_fetch"
        end
      end
    end
  end
end

File: crates/agrr-domain/src/weather_data/gateways/internal_weather_fetch_start_gateway.rs
//! Ruby: `Domain::WeatherData::Gateways::InternalWeatherFetchStartGateway`

use crate::weather_data::dtos::InternalWeatherFetchStartOutput;

/// Ruby snapshot struct
#[derive(Debug, Clone)]
pub struct WeatherFetchFarmSnapshot {
    pub farm_id: i64,
    pub weather_data_status: String,
    pub weather_data_count: Option<i32>,
    pub total_blocks: i32,
}

#[derive(Debug, Clone)]
pub enum StartInternalWeatherFetchResult {
    FarmNotFound,
    Completed(WeatherFetchFarmSnapshot),
    Started(WeatherFetchFarmSnapshot),
    NeedsFetch(WeatherFetchFarmSnapshot),
    Failed(String),
}

/// Ruby: `Domain::WeatherData::Gateways::InternalWeatherFetchStartGateway`
pub trait InternalWeatherFetchStartGateway: Send + Sync {
    fn start_internal_weather_data_fetch(
        &self,
        farm_id: &str,
    ) -> StartInternalWeatherFetchResult;
}

/// Started farm snapshot after triggering fetch (farm interactor result).
#[derive(Debug, Clone)]
pub struct StartedFarmWeatherFetchSnapshot {
    pub weather_data_status: String,
    pub weather_data_total_years: i32,
}

/// Ruby: farm gateway slice for weather fetch start.
pub trait StartFarmWeatherDataFetchPort: Send + Sync {
    fn call(
        &self,
        farm_id: i64,
        as_of: time::Date,
    ) -> Option<StartedFarmWeatherFetchSnapshot>;
}
